code/parachain/frame/cosmwasm/src/lib.rs

```diff
@@ -507,6 +555,58 @@ pub mod pallet {
 			Ok((contract, contract_info))
 		}
 
+		/// Setup for a contract migration.
+		/// This is called prior to calling the `migrate` export of the contract.
+		pub(crate) fn do_migrate_phase1(
+			migrator: &AccountIdOf<T>,
+			contract: &AccountIdOf<T>,
+			new_code_id: CosmwasmCodeId,
+		) -> Result<ContractInfoOf<T>, Error<T>> {
+			CodeIdToInfo::<T>::try_mutate_exists(new_code_id, |entry| -> Result<(), Error<T>> {
+				let code_info = entry.as_mut().ok_or(Error::<T>::CodeNotFound)?;
+				code_info.refcount =
+					code_info.refcount.checked_add(1).ok_or(Error::<T>::RefcountOverflow)?;
+				Ok(())
+			})?;
+
+			let (contract_info, code_id) = ContractToInfo::<T>::try_mutate(
+				contract,
+				|entry| -> Result<(ContractInfoOf<T>, u64), Error<T>> {
+					let info = entry.as_mut().ok_or(Error::<T>::ContractNotFound)?;
+					ensure!(info.admin.as_ref() == Some(migrator), Error::<T>::NotAuthorized);
+					let old_code_id = info.code_id;
+					info.code_id = new_code_id;
+					Ok((info.clone(), old_code_id))
+				},
+			)?;
+
+			CodeIdToInfo::<T>::try_mutate_exists(code_id, |entry| -> Result<(), Error<T>> {
+				let code_info = entry.as_mut().ok_or(Error::<T>::CodeNotFound)?;
+				code_info.refcount =
+					code_info.refcount.checked_sub(1).ok_or(Error::<T>::RefcountOverflow)?;
+				if code_info.refcount == 0 {
+					// Code is unused after this point, so it can be removed
+					*entry = None;
+					let code = PristineCode::<T>::try_get(code_id)
+						.map_err(|_| Error::<T>::CodeNotFound)?;
+					let deposit = code.len().saturating_mul(T::CodeStorageByteDeposit::get() as _);
+					let _ = T::NativeAsset::unreserve(migrator, deposit.saturated_into());
+					let code_hash = T::Hashing::hash(&code);
+					PristineCode::<T>::remove(code_id);
+					InstrumentedCode::<T>::remove(code_id);
+					CodeHashToId::<T>::remove(code_hash);
+				}
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::<T>::Migrated {
+				contract: contract.clone(),
+				to: new_code_id,
+			});
+
+			Ok(contract_info)
+		}
```

Comment:
Nit: this pattern allows for incorrectly migrating a contract. Instead we could use a Migrator struct, using Typestate to ensure that `migrate_phase1` is called before a contract is migrated.

Reply:
In general, let's avoid patterns where foo_1 needs to be called before foo, and encode that logic in the type system.